diff --git a/node_modules/@nestjs/graphql/dist/services/gql-subscription.service.js b/node_modules/@nestjs/graphql/dist/services/gql-subscription.service.js
index ab516b5..98a21c2 100644
--- a/node_modules/@nestjs/graphql/dist/services/gql-subscription.service.js
+++ b/node_modules/@nestjs/graphql/dist/services/gql-subscription.service.js
@@ -57,6 +57,9 @@ class GqlSubscriptionService {
                 !protocols.includes(graphql_ws_1.GRAPHQL_TRANSPORT_WS_PROTOCOL) // graphql-ws subprotocol
                 ? this.subTransWs
                 : this.wss;
+            if (req.url.startsWith('/_next')) {
+                return;
+            }
             wss.handleUpgrade(req, socket, head, (ws) => {
                 wss.emit('connection', ws, req);
             });
